Add `makeself` backend (#20360)
pantsbuild/pants: docs/docs/shell/self-extractable-archives.mdx

Reviewer: Just observing that this is scoped within the "shell" docs, which is nice. Do you think the whole backend could potentially be scoped within the shell one? Or is that unnecessary? Like `pants.backend.experimental.shell.makeself`

Author: I believe with some hacks it can be used without shell scripts, e.g. python or any other scripting language that supports shebangs. Most probably you won't do this, but I think we can keep it in it's own space. 

We can come back to this question once we decide to stabilize the backend

```bash
dist/arc.run
```
```
Verifying archive integrity...  100%   MD5 checksums are OK. All good.
Uncompressing arc.run  100%  
hello pants
```

The built archive supports a bunch of parameters, you can inspect them manually:

```bash
dist/arc.run --help
```

Or refer to the [`makeself`](https://github.com/megastep/makeself) documentation.

## Bundling multiple files

You can bundle multiple shell scripts using
[`files`](../../reference/targets/makeself_archive.mdx#files) field:

```python tab={"label":"BUILD"}
shell_sources(name="src")

makeself_archive(
    name="arc",
    files=["lib.sh:src", "entrypoint.sh:src"],
    startup_script=["./entrypoint.sh"],
)
```

```bash tab={"label":"entrypoint.sh"} 
#!/bin/bash

. lib.sh
echo $@ "one two three" | first_column
```

```bash tab={"label":"lib.sh"} 
#!/bin/bash

function first_column {
    awk '{print $1}'
}
```

Notice that we need to use a relative path to the `./entrypoint.sh`.

```bash
pants package :arc && dist/arc.run
```
```
[INFO] Wrote dist/arc.run
Built Makeself binary: arc.run
Verifying archive integrity...  100%   MD5 checksums are OK. All good.
Uncompressing arc.run  100%  
one
```

> To pass the arguments to the startup_script use `--`:

Reviewer: ```suggestion
To pass the arguments to the `startup_script` use `--`:
```

```bash
pants package :arc && dist/arc.run -- zero
```
```
[INFO] Wrote dist/arc.run
Built Makeself binary: arc.run
Verifying archive integrity...  100%   MD5 checksums are OK. All good.
Uncompressing arc.run  100%  
zero
```

## `pants run`

> Instead of packaging and running `makeself_archive` manually, you can use `run` goal instead:

Reviewer: ```suggestion
Instead of packaging and running `makeself_archive` manually, you can use the `run` goal instead:
```

```bash
pants run :arc
```
```
Verifying archive integrity...  100%   MD5 checksums are OK. All good.
Uncompressing arc.run  100%  
one
```

To pass the arguments through the `pants run` goal you need `--`, then you need

> another `--` to pass arguments to archive `startup_script`, so you end up with

Reviewer: ```suggestion
another `--` to pass arguments to the archive's `startup_script`, so you end up with
```

]

[python]
interpreter_constraints = ["CPython==3.12.*"]
```

Now define the `pex_binary` and add it to the makeself archive via the
`packages` field:

```python tab={"label":"BUILD"}
python_sources(name="py")
pex_binary(
    name="upper",
    entry_point="upper.py",
)
shell_sources(name="sh")
makeself_archive(
    name="arc",
    files=["lib.sh:sh", "entrypoint.sh:sh"],
    packages=[":upper"],
    startup_script=["./entrypoint.sh"],
)
```

```python tab={"label":"upper.py"}
print(input().upper())
```

```bash tab={"label":"entrypoint.sh"} 
#!/bin/bash

. lib.sh
echo $@ "one two three" | first_column | ./upper.pex
```

```bash tab={"label":"lib.sh"} 
#!/bin/bash

function first_column {
    awk '{print $1}'
}
```

```bash
pants run :arc -- -q -- zero
```
```
/usr/bin/env: ‘python3.12’: No such file or directory
```

Oops! This happened because `pants run` is running in isolated environment, so
we have to explicitly tell pants that we want to access the python interpreter
and a couple of binaries used by pex:

```python title="BUILD"
...
makeself_archive(
    name="arc",
    files=["lib.sh:sh", "entrypoint.sh:sh"],
    packages=[":upper"],
    startup_script=["./entrypoint.sh"],

>     tools=["python3.12", "grep", "sort"],

Reviewer: Hm, just observing that these tools seem to be used for running... but I imagine they're not required for packaging? Potentially something to iterate on?

Author: This is specifically about the pex inside. Since the `run` goal doesn't know there is a pex inside a makeself archive, you need to explicitly add python, grep and sort which are used inside the pex launcher, but not required for makeself launcher.

This is not needed for shell-only makeself archives.

Reviewer: Yeah, I understand it's just for this particular archive... but I think the tools are only used when running the makeself archive using `pants run`? That is, even if this had no tools specified `tools=[]`, I think the `pants package ...` goal would still work.

No need to change, but just observing that there's a distinction between "tools required to package" and "tools required to run", and overlaying them may lead to spurious cache invalidations (etc.).

Author: Ah, yes, this is correct

I could split it into `execution_dependencies` and `runnable_dependencies` similar to `adhoc_tool`
